### PR TITLE
Store a GenericSignature in GenericEnvironment

### PR DESCRIPTION
**Pull request by huonw**
Almost every place that manipulates a GenericEnvironment also has a GenericSignature, so packaging them together makes life easier.

### CODE REVIEW

**Comment by huonw** [comment]
@swift-ci Please test.

**Comment by huonw** [comment]
cc @slavapestov and @DougGregor 

Most of this is straight forward but I don't understand the second commit (40c98d4 at the moment)... this change hits the assertion, but nothing seems to break horribly when it's just removed.

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -1467,13 +1458,7 @@ class ExtensionDecl final : public Decl, public DeclContext,
...
 
   /// Retrieve the generic signature for this extension.
-  GenericSignature *getGenericSignature() const { return GenericSig; }
-
-  /// Set the generic signature of this extension.
-  void setGenericSignature(GenericSignature *sig) {
-    assert(!GenericSig && "Already have generic signature");
-    GenericSig = sig;
-  }
+  GenericSignature *getGenericSignature() const { return GenericEnv ? GenericEnv->getGenericSignature() : nullptr; }
```
Comment (slavapestov) [inline include/swift/AST/Decl.h]: Line length

File: include/swift/AST/TypeRepr.h [hunk 2]
```diff
@@ -380,19 +380,13 @@ class FunctionTypeRepr : public TypeRepr {
...
 
   GenericParamList *getGenericParams() const { return GenericParams; }
   GenericEnvironment *getGenericEnvironment() const { return GenericEnv; }
-  GenericSignature *getGenericSignature() const { return GenericSig; }
-
-  void setGenericSignature(GenericSignature *genericSig) {
-    assert(GenericSig == nullptr);
-    GenericSig = genericSig;
-  }
+  GenericSignature *getGenericSignature() const { return GenericEnv ? GenericEnv->getGenericSignature() : nullptr; }
```
Comment (slavapestov) [inline include/swift/AST/TypeRepr.h]: Line length

File: include/swift/SIL/TypeSubstCloner.h [hunk 3]
```diff
@@ -41,8 +41,7 @@ class TypeSubstCloner : public SILClonerWithScopes<ImplClass> {
 
   void computeSubsMap() {
     if (auto *env = Original.getGenericEnvironment()) {
-      auto sig = Original.getLoweredFunctionType()->getGenericSignature();
-      SubsMap = env->getSubstitutionMap(SwiftMod, sig, ApplySubs);
+      SubsMap = env->getSubstitutionMap(SwiftMod, ApplySubs);
```
Comment (slavapestov) [inline include/swift/SIL/TypeSubstCloner.h]: Somewhere in SILFunction can you assert that F->getGenericEnvironment()->getGenericSignature()->getCanonicalSignature() == F->getLoweredFunctionType()->getGenericSignature()?

File: lib/Sema/TypeCheckDecl.cpp [hunk 4]
```diff
@@ -7529,7 +7529,8 @@ checkExtensionGenericParams(TypeChecker &tc, ExtensionDecl *ext,
 
   // Validate the generic type signature.
   auto *sig = tc.validateGenericSignature(genericParams,
-                                          ext->getDeclContext(), parentSig,
+                                          ext->getDeclContext(),
+                                          parentSig,
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: Best to avoid no-op changes like this

**Review by slavapestov — COMMENTED** [review]
Looks pretty sensible, other than the FIXME. I'll explain that part on Monday.

**Comment by dabrahams** [comment]
@swift-ci Please test OS X platform

**Comment by dabrahams** [comment]
@swift-ci Please test OS X platform

**Comment by DougGregor** [comment]
Looks like the macOS failure is because a compiler-crasher (compiler_crashers/28474-unreachable-executed-at-swift-lib-ast-type-cpp-1325.swift) is no longer crashing...

**Comment by DougGregor** [comment]
@swift-ci please smoke test OSX

**Comment by ematejska** [comment]
@swift-ci please smoke test OS X

**Comment by huonw** [comment]
I added a few commits, which hopefully address the review, and fixes the FIXME commit. I've left them separate for now, but will clean up the history before merging.

**Comment by huonw** [comment]
(Looking at the unexpected pass now.)

**Comment by DougGregor** [comment]
The new bits look reasonable to me

**Comment by huonw** [comment]
@swift-ci Please test.

**Comment by huonw** [comment]
@swift-ci Please test.

**Comment by huonw** [comment]
@swift-ci Please smoke test and merge.

**Pull request merged**